midp/gfx.js

```diff
@@ -112,8 +112,35 @@ var currentlyFocusedTextEditor;
         return 1;
     };
 
+    var hasNewFrame = true;
+    var ctxToExecuteOnNewFrame;
+    function gotNewFrame(timestamp) {
+        if (ctxToExecuteOnNewFrame) {
+            var ctx = ctxToExecuteOnNewFrame;
+            ctxToExecuteOnNewFrame = null;
+            window.requestAnimationFrame(gotNewFrame);
+            ctx.execute();
+        } else {
+            hasNewFrame = true;
+        }
+    }
 
-    addUnimplementedNative("com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V");
+    Native["com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V"] = function() {
+        // This is a no-op: The foreground display gets drawn directly
+        // to the screen.
+        //
+        // However, we sometimes "draw" more frequently than the
+        // screen can update. By calling `requestAnimationFrame` here,
+        // we ensure that whatever we just drew will stick around for
+        // one frame
+        if (hasNewFrame) {
+            hasNewFrame = false;
+            window.requestAnimationFrame(gotNewFrame);
+        } else {
+            ctxToExecuteOnNewFrame = $.ctx;
+            $.pause(asyncImplStringAsync);
```

Comment:
> I believe that RepaintEventProducer.process should only ever be called from RepaintEventProducer.serviceRepaints after this check:

It's unlikely, but there could be two isolates each of which would have there own event queues.

As for the other thing about waiting. This seems to be a wash. Seth, one of the guys familiar with graphics here said to try both, but we'll never seen anything updated out of band with a requestAnimationFrame. As for the code in this patch...you'll always be waiting for the requestAnimationFrame cb so I don't see how it is any different than my suggestion above. e.g. if hasNewFrame=true you still do window.requestAnimationFrame.